client/container_create.go

```diff
@@ -61,3 +60,20 @@ func (cli *Client) ContainerCreate(ctx context.Context, config *container.Config
 	err = json.NewDecoder(serverResp.body).Decode(&response)
 	return response, err
 }
+
+// formatPlatform returns a formatted string representing platform (e.g. linux/arm/v7).
+//
+// Similar to containerd's platforms.Format(), but returns an empty string
+// instead of "unknown" if no OS is present: https://github.com/containerd/containerd/blob/v1.5.2/platforms/platforms.go#L243-L263
+func formatPlatform(platform *specs.Platform) string {
```

Comment:
Seems like we really want `path.Join(platform.OS, platform.Arch....)`
This will handle empty values as desired, even for empty values.
https://play.golang.org/p/ztICsyudOrB

Not sure if you want to validate that is always arch is set if variant is also set.

Reply:
Oh, nice! TIL `path.Join` skips the empty values (I assumed it would keep those in). Yes, I like that.

> Not sure if you want to validate that is always arch is set if variant is also set.

Yeah, perhaps stop at the first empty value (in which case `path.Join` would not be needed 🤔). We could keep it "simple" (for now), and just let the daemon handle this; wdyt?

I was writing down notes on serializing / deserialising and normalising of oci.Platform (to start a discussion on the OCI spec), because currently it looks like there's no standard approach anywhere.


So add to the "silliness"; I was looking at the CLI code YesterdayI: https://github.com/docker/cli/blob/v20.10.7/cli/command/container/create.go#L244-L249


```go
	if opts.platform != "" && versions.GreaterThanOrEqualTo(dockerCli.Client().ClientVersion(), "1.41") {
		p, err := platforms.Parse(opts.platform)
		if err != nil {
			return nil, errors.Wrap(err, "error parsing specified platform")
		}
		platform = &p
```

So, we:

1. pass a string representation of platform (through the `--platform` flag)
2. parse it on the CLI to convert it to an `oci.Platform{}`
3. pass it to the client as argument
4. convert it to a string
5. send it over the API as a string
6. parse the string and convert it to an `oci.Platform{}`
7. pass it to the backend as argument

Looking at that, I think we should consider to either have the client (and cli) to not bother with `oci.Platform{}` at all (pass a string, and let the daemon deal with it), or make the API accept a `oci.Platform{}` (or equivalent) in the request.